### PR TITLE
Fix Duration.between throwing an exception in some scenarios

### PR DESCRIPTION
**Pull request by alzimmermsft**
# Description

Flips the arguments when calling `Duration.between` and negates the value when logging cache refreshes in `AccessTokenCache`. This is done as some `TokenCredential`s use a static `OffsetDateTime.MAX` as the expiration time for the token, and in some cases using `OffsetDateTime.MAX` as the second argument to `Duration.between` can cause an invalid year range exception to be thrown. The exception happens due to the implementation of `Duration.between` in some versions of Java changes the time zone of the second argument to match the first, resulting in `OffsetDateTime.MAX` needing to rollover its second, minute, hour, day, month, and year values due to the time zone change, causing its `999,999` year value (which is the max allowed by OffsetDateTime) to become `1,000,000`. Flipping the arguments to `Duration.between` resolves this as the first argument doesn't have its time zone mutated, meaning `OffsetDateTime.MAX` won't be mutated, and should be safe always as `OffsetDateTime.MAX` uses the most minimal time zone offset allowed, meaning any Temporal near the max year limit won't need to rollover values. This does force the `Duration.between` to be `.negated()` as the values have flipped.

There are a few other small changes in this PR to have `TokenCredential` subtypes implement the synchronous token retrieval API instead of relying on the default implementation in `TokenCredential`.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/implementation/AccessTokenCache.java [hunk 1]
```diff
@@ -290,7 +290,7 @@ private static LoggingEventBuilder buildTokenRefreshLog(LogLevel level, AccessTo
             return logBuilder;
         }
 
-        Duration tte = Duration.between(now, cache.getExpiresAt());
+        Duration tte = Duration.between(cache.getExpiresAt(), now).negated();
```
Comment (billwert) [inline sdk/core/azure-core/src/main/java/com/azure/core/implementation/AccessTokenCache.java:293]: This probably deserves a comment explaining the issue and the fix.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/EventHubSharedKeyCredential.java [hunk 2]
```diff
@@ -35,12 +35,12 @@
  * <p>
  * The shared access key can be obtained by creating a <i>shared access policy</i> for the Event Hubs namespace or for
  * a specific Event Hub instance. See
- * <a href="https://docs.microsoft.com/en-us/azure/event-hubs/
+ * <a href="https://docs.microsoft.com/azure/event-hubs/
```
Comment (billwert) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/EventHubSharedKeyCredential.java:38]:
```suggestion
 * <a href="https://learn.microsoft.com/azure/event-hubs/
```

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/EventHubSharedKeyCredential.java [hunk 3]
```diff
@@ -35,12 +35,12 @@
...
  * authorize-access-shared-access-signature#shared-access-authorization-policies">Shared access authorization policies
  * </a> for more information.
  * </p>
  *
- * @see <a href="https://docs.microsoft.com/en-us/azure/event-hubs/authorize-access-shared-access-signature">Authorize
+ * @see <a href="https://docs.microsoft.com/azure/event-hubs/authorize-access-shared-access-signature">Authorize
```
Comment (billwert) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/EventHubSharedKeyCredential.java:43]:
```suggestion
 * @see <a href="https://learn.microsoft.com/azure/event-hubs/authorize-access-shared-access-signature">Authorize
```

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/authentication/AccessTokenCacheImpl.java [hunk 4]
```diff
@@ -287,7 +287,7 @@ private static LoggingEventBuilder buildTokenRefreshLog(LogLevel level, AccessTo
             return logBuilder;
         }
 
-        Duration tte = Duration.between(now, cache.getExpiresAt());
+        Duration tte = Duration.between(cache.getExpiresAt(), now).negated();
```
Comment (billwert) [inline sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/authentication/AccessTokenCacheImpl.java:290]: Comment here also.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/MSIToken.java [hunk 5]
```diff
@@ -90,12 +84,13 @@ private static Long parseToEpochSeconds(String expiresOn, String expiresIn) {
         // expiresOn = timestamp of refresh expressed as seconds since epoch.
 
         // if we have an expiresOn, we'll use it. Otherwise, we use expiresIn.
-        String dateToParse = CoreUtils.isNullOrEmpty(expiresOn) ? expiresIn : expiresOn;
+        boolean isExpiresOn = !CoreUtils.isNullOrEmpty(expiresOn);
+        String dateToParse = !isExpiresOn ? expiresIn : expiresOn;
```
Comment (billwert) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/MSIToken.java:88]:
```suggestion
        String dateToParse = isExpiresOn ? expiresOn: expiresIn;
```

There's enough negativity in the world.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusSharedKeyCredential.java [hunk 6]
```diff
@@ -36,12 +36,12 @@
  * <p>
  * The shared access key can be obtained by creating a <i>shared access policy</i> for the Service Bus namespace or for
  * a specific Service Bus instance. See
- * <a href="https://docs.microsoft.com/en-us/azure/event-hubs/
+ * <a href="https://docs.microsoft.com/azure/event-hubs/
```
Comment (billwert) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusSharedKeyCredential.java:39]:
```suggestion
 * <a href="https://learn.microsoft.com/azure/event-hubs/
```

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusSharedKeyCredential.java [hunk 7]
```diff
@@ -36,12 +36,12 @@
...
  * authorize-access-shared-access-signature#shared-access-authorization-policies">Shared access authorization policies
  * </a> for more information.
  * </p>
  *
- * @see <a href="https://docs.microsoft.com/en-us/azure/event-hubs/authorize-access-shared-access-signature">Authorize
+ * @see <a href="https://docs.microsoft.com/azure/event-hubs/authorize-access-shared-access-signature">Authorize
```
Comment (billwert) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/ServiceBusSharedKeyCredential.java:44]:
```suggestion
 * @see <a href="https://learn.microsoft.com/azure/event-hubs/authorize-access-shared-access-signature">Authorize
```

**Review by billwert — CHANGES_REQUESTED** [review]
I worry that we're leaving traps for future developers by hiding the negation in the cache implementations. Couple ideas:
1) Util method in Core for `Duration.between` with the arguments swapped and negated. (Maybe ban the `Duration.between` and make the whole codebase use the util method. Do we have a mechanism for that?)
2) Maybe this is just an `AccessToken` problem in practice, so put the helper on `AccessToken` to get the remaining life. `validDuration`, `remainingDuration`, `timeToLive`, etc?

You suggested elsewhere just using some absurdly large `OffsetDateTime` instead - I think that's fine also but would still like a constant somewhere for it and some sort of linter check to prevent mistakes in the future.

**Review by billwert — APPROVED** [review]

**Comment by alzimmermsft** [comment]
Failing tests and pipelines are due to changes other than this PR

**Pull request merged**